src/bpf/bpfctl.c

```diff
@@ -0,0 +1,220 @@
+                if (r < 0)
+                        return r;
+
+                log_info(" map_name: %s, map_type: %lu, memlock: %s", p.name, p.type, p.memlock);
+        } else {
+                static const JsonDispatch dispatch_table[] = {
+                        { "NAME",  JSON_VARIANT_STRING, json_dispatch_string, offsetof(struct bpf_prog_data, name),  JSON_MANDATORY },
+                        { "TYPE", JSON_VARIANT_UNSIGNED, json_dispatch_uint64, offsetof(struct bpf_prog_data, type), JSON_MANDATORY},
+                        { "MEMLOCK", JSON_VARIANT_STRING, json_dispatch_string, offsetof(struct bpf_prog_data, memlock), JSON_MANDATORY},
+                        { "TAG", JSON_VARIANT_STRING, json_dispatch_string, offsetof(struct bpf_prog_data, tag), JSON_MANDATORY},
+                        { "MAP_NAMES", JSON_VARIANT_STRING, json_dispatch_string, offsetof(struct bpf_prog_data, map_names), 0},
+                        { "RUN_TIME_NS", JSON_VARIANT_UNSIGNED, json_dispatch_uint64, offsetof(struct bpf_prog_data, run_time_ns), JSON_MANDATORY},
+                        { "RUN_COUNT", JSON_VARIANT_UNSIGNED, json_dispatch_uint64, offsetof(struct bpf_prog_data, run_cnt), JSON_MANDATORY},
+                        {}
+                };
+                struct bpf_prog_data p = {
+                        .map_names = NULL,
+                };
+
+                r = json_dispatch(v, dispatch_table, NULL, 0, &p);
+                if (r < 0)
+                        return r;
+
+                log_info(" prog_name: %s, prog_type: %lu, memlock: %s, prog_tag: %s, prog_map_names: %s, run_time_ns: %lu, run_cnt: %lu", p.name, p.type, p.memlock, p.tag, p.map_names, p.run_time_ns, p.run_cnt);
+        }
+
+        return 0;
+}
+
+static int query_bpf_progs_and_maps(void) {
+        _cleanup_(sd_event_unrefp) sd_event *event = NULL;
+        _cleanup_(varlink_unrefp) Varlink *vl = NULL;
+        int r;
+
+        r = sd_event_default(&event);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get event loop: %m");
+
+        r = varlink_connect_address(&vl, "/run/systemd/bpf/io.systemd.Bpf");
+        if (r < 0)
+                return log_debug_errno(r, "Unable to connect to /run/systemd/bpf/io.systemd.Bpf: %m");
+
+        r = varlink_attach_event(vl, event, SD_EVENT_PRIORITY_NORMAL);
+        if (r < 0)
+                return log_error_errno(r, "Failed to attach varlink connection to event loop: %m");
+
+        r = varlink_bind_reply(vl, query_bpf_progs_and_maps_reply);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to bind reply callback: %m");
+        char method[] = "io.systemd.BpfProgsAndMaps.GetBpfProgsAndMaps";
+        r = varlink_observe(vl, method, NULL);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to invoke varlink method: %m");
+
+        r = sd_event_loop(event);
+        if (r < 0)
+                return log_error_errno(r, "Failed to run event loop: %m");
```

Comment:
if there's no strong reason to use an event loop here (i.e. that you have multiple events to care for) then I'd just lop around varlink_process() here

Reply:
Agreed - ideally using varlink_collect here will do the trick.